repo/tuf_on_ci/client.py

```diff
@@ -5,74 +5,110 @@
 from tempfile import TemporaryDirectory
+from urllib import request
 
 import click
+from tuf.api.exceptions import ExpiredMetadataError
+from tuf.api.metadata import Metadata
 from tuf.ngclient import Updater
 
 
+def expiry_check(dir: str, role: str, timestamp: int):
+    ref_time = datetime.fromtimestamp(timestamp)
+    md = Metadata.from_file(os.path.join(dir, f"{role}.json"))
+    expiry = md.signed.expires
+    if ref_time > expiry:
+        sys.exit(f"Error: {role} expires {expiry} (expected valid at {ref_time})")
+    print(f"Role {role} is valid on {ref_time}: OK")
+
+
 @click.command()
 @click.option("-v", "--verbose", count=True, default=0)
 @click.option("-m", "--metadata-url", type=str, required=True)
 @click.option("-a", "--artifact-url", type=str, required=True)
+@click.option("-u", "--update-base-url", type=str)
 @click.option("-e", "--expected-artifact", type=str)
-@click.option(
-    "-c",
-    "--metadata-cache",
-    type=str,
-    help="Optional client cache dir that client should use",
-)
+@click.option("--compare-source/--no-compare-source", default=True)
+@click.option("-t", "--time", type=int)
+@click.option("-o", "--offline-time", type=int)
+@click.option("-d", "--metadata-dir", type=str)
 def client(
     verbose: int,
     metadata_url: str,
     artifact_url: str,
+    update_base_url: str | None,
     expected_artifact: str | None,
-    metadata_cache: str | None,
+    compare_source: bool,
+    time: int | None,
+    offline_time: int | None,
+    metadata_dir: str | None,
 ) -> None:
-    """Test client for tuf-on-ci
-
-    Client expects to be run in tuf-on-ci repository (current metadata
-    in the sources will be considered the expected expected metadata the
-    client should receive from remote)."""
+    """Test client for tuf-on-ci"""
 
     logging.basicConfig(level=logging.WARNING - verbose * 10)
 
     with TemporaryDirectory() as client_dir:
-        metadata_dir = os.path.join(client_dir, "metadata")
+        if metadata_dir is None:
+            metadata_dir = os.path.join(client_dir, "metadata")
         artifact_dir = os.path.join(client_dir, "artifacts")
-        os.mkdir(metadata_dir)
+        os.makedirs(metadata_dir, exist_ok=True)
         os.mkdir(artifact_dir)
 
-        if metadata_cache:
-            # pre-populate the client cache with given directory
-            for fname in os.listdir(metadata_cache):
-                fpath = os.path.join(metadata_cache, fname)
-                if os.path.isfile(fpath):
-                    shutil.copy(fpath, os.path.join(metadata_dir, fname))
-        else:
-            # pre-populate the client cache with first root version from source
-            root = "metadata/root_history/1.root.json"
-            shutil.copy(root, os.path.join(metadata_dir, "root.json"))
-
-        # For now, just confirm we can get top level metadata from remote
+        # initialize client with a root.json from metadata_url
+        root_url = f"{metadata_url}/1.root.json"
```

Comment:
If we want to support an option to start at a given version of the root (like 5 for sigstore), how about an option that specifies the starting version that is used here (could be simple implement as the default argument can be `1`), which can be used here? It may not be as complicated as an new option for the starting root url?

Reply:
I implemented this using "root.json" from working directory: this way the workflow can provide the file in whatever way they want